Adds an alternative method to compute eta_function in CustomSD to deal with large memory-lengths

emile-cochin opened the pull request:
# Summary

This adds an `alt_integrator` parameter to `PowerLawSD`/`CustomSD` and to `eta_function`/`correlation_2d_integral` that dictates whether or not to use an alternative method to numerically integrate `eta_function` (splitting the integrals into oscillating terms and using cos/sin weighted integrals instead as described more precisely in #150). Also adds corresponding coverage tests. It adds a new physics test `bath_correlations_test.py` which compares exact values of these 2d integrals (for an Ohmic SD with exponential cutoff at 0 temperature) to both versions of the integrators, and for short and long memory-lengths. More specifically, this test expects either a sufficiently good agreement or at least an `IntegrationWarning` being raised if the results don't match. For now it passes in about 2s with the default integrator issuing a warning for long memory-lengths.

This is a draft since there are a few choices to discuss before this is potentially publishable. I'm notably thinking about:
- Whether giving the choice through this additional parameter is the way to go (or if its name is sufficiently clear).
- What choice of $\tilde{\omega}$ (`omega_tilde`) is best suited for each $\tau$ (`tau`). For now the function implemented is `omega_tilde = omega_cutoff/int(2+omega_cutoff*tau/10)` which seemed to give convincing results for a large number of spectral densities and parameter choices.
It is heuristically motivated since the idea is to have a smaller first interval for larger $\tau$ to limit the number of oscillations that the non-specialized integrator has to deal with. Still, it looks a bit arbitrary and it might be nice to let the user be able to play around with this for their particular use case (in which case one could pass a function to `alt_integrator` or alternatively rewrite all of this to pass an `integrator_options` dictionary instead).
- If there should be a way to visualize the results of `correlation_2d_integral` in the `helpers` submodule and, if so, if it should let the user plot any correlation or specifically the ones as used in the `influence_matrix` function.
- If I should add an example of how to use this feature or if it is small enough not to be necessary.

# Pull Request Check List

This checklist serves as a guide for contributors and maintainers to assess
whether a contribution can be merged into the main branch and thus serves
to guarantee the quality of the package. Please read the
[contribution guideline](https://github.com/tempoCollaboration/OQuPy/blob/main/CONTRIBUTING.md).

Before you submit a pull request, please go through this checklist once
more, as it will decrease the number of unnecessary review cycles.
However, this list is there to *help* all contributors to produce high
quality code, not to deter you from contributing. If you can't tick some of the
boxes, feel free to submit the pull request anyway and report about it in the
pull request message. Also, some of the boxes might not apply to the specific
type of your contribution.

* [x] The contribution has been discussed and agreed on in the [Issue section](https://github.com/tempoCollaboration/OQuPy/issues).
* [x] Code contributions do its best to follow [the zen of python](https://www.python.org/dev/peps/pep-0020/).
* [x] The automated test are all positive:
  - [x] `tox -e py36` (to run `pytest`) the code tests.
  - [x] `tox -e style` (to run `pylint`) the [code style](https://www.python.org/dev/peps/pep-0008/) tests.
* [x] Added test for changed/added code.
* [x] The documentation has been updated:
  - [x] docstring for all new functions/methods/classes/modules.
  - [x] consistent style of all docstrings.
  - [ ] for new modules: `/docs/pages/modules.rst` has been updated.
  - [ ] for new functionality: `/docs/pages/api.rst` has been updated.
  - [ ] for new functionality: tutorials and examples have been updated.

emile-cochin committed:
Adds an alternative method to compute eta_function in CustomSD

Adds an alt_integrator parameter to PowerLawSD/CustomSD and to
eta_function/correlation_2d_integral that dictates whether or not to use
an alternative method to numerically integrate eta_function (splitting
the integrals into oscillating terms and using cos/sin weighted
integrals instead). Also adds corresponding coverage tests and physics
tests comparing exact values of these integrals to both versions of the
integrator (asking for at least an IntegrationWarning if the results
don't match).

emile-cochin committed:
Added a way to change the first cutoff when using alt_integrator in CustomSD

piperfw commented on oqupy/bath_correlations.py:
```diff
@@ -390,6 +390,21 @@ class CustomSD(BaseCorrelations):
         Whether or not to use a sine/cosine weighted version of the
```
Suggest small revision to mention scipy,

"Whether to use an alternative integration scheme leveraging sine/cosine weighted versions of scipy.integrate.quad to handle rapid oscillations. This may improve numerical accuracy and stability, especially for long memory times.
[Blank line]
The integral...
"

piperfw commented on oqupy/bath_correlations.py:
```diff
@@ -442,6 +459,24 @@ def __init__(
         assert isinstance(alt_integrator, bool)
         self._alt_integrator = alt_integrator
 
+        # create the first_cutoff function and input check for num_oscillations
```
Do we want to assert that, if num_oscillations is provided, alt_integrator should be true? Or at least provide a warning that it will be discarded when alt_integrator is False?

emile-cochin replied:
I added a warning [L478](https://github.com/tempoCollaboration/OQuPy/pull/151/changes/3f186bde3dd0c5042c79da711575edf4ebace2d3#diff-e6bdfb99eb07a7b5a46647a13d1b42b9c8d4bd1a2b55d91075dc31393c6c7e87R478).

piperfw commented on oqupy/bath_correlations.py:
```diff
@@ -851,6 +886,21 @@ class PowerLawSD(CustomSD):
         Whether or not to use a sine/cosine weighted version of the
         integrals that could be better suited if you're encountering
         numerical difficulties, especially for long memory times.
+        The integral computed in `eta_function` is cut in two: one
```
Make consistent with above

piperfw commented on oqupy/bath_correlations.py:
```diff
@@ -362,6 +386,10 @@ class CustomSD(BaseCorrelations):
         ``'gaussian'``}
     temperature: float
         The environment's temperature.
+    alt_integrator: bool
```
I think in docstrings other than CustomSD (or whatever you deem the 'parent' one to be) we can either a) copy the entire docstring or b) (probably neater) reference CustomSD, e.g.

Whether to use an alternative integration scheme leveraging sine/cosine weighted versions of scipy.integrate.quad to handle rapid oscillations. See...

piperfw reviewed (commented):
Some comments on docstrings relevant to ongoing Issue discussion

piperfw commented:
Great input on streamlining @gefux, I agree simplifying the logic we have here would be beneficial. Just on your points

1.  There is one known issue that can affect us, that the weighted integrals [do not handle small wvars incorrectly](https://github.com/scipy/scipy/issues/12684). Indeed Emile mentions problems when $\tau$ is small in #150. In fact, I think we need to check for and protect against this case whether we keep or replace the previous method. The interface simplification from having a single method is appealing. 

A related question, @emile-cochin have you done any performance testing? Not that we prioritise fast code, but it would be interesting to know whether the gain from using the weighted version outweighs the overhead from replacing two calls to `scipy.quad` with (I think) six calls, for every `eta` integral.

2. Two separate versions sounds good unless Emile particularly wants to work on unifying them.

3. A dictionary would work well, subject to 1.: if we keep both methods, then the flexibility Emile included by allowing method parameters to override the value of `alt_integrator` passed to the `customSD` constructor may be useful e.g. to handle small `tau`?

emile-cochin committed:
added a parameters dictionary for integration parameters in CustomSD

Put previous `epsrel`, `subdiv_limit` parameters together with
`alt_integrator` and `num_oscillations` in a single parameter dictionary
as input for `CustomSD` (and necessary modifications resulting from this
change in tests/) which defaults to `INTEGRATION_PARAMS` in `config.py`.

Also adds a small section in the parameters tutorials about the
alternative integrator.

emile-cochin commented on oqupy/bath_correlations.py:
```diff
@@ -583,6 +601,56 @@ def integrand(w):
             integral = integral.real
         return integral
 
+    def _eta_function_alt(self,
```
I added this function for computation of `eta_function` with the new method so that the main function is less cluttered.

gefux committed:
Simplify eta_function integral computation using weighted quadrature

* untangle matsubara computation from default eta_function computation

* simplify the weighted quadrature parameter to just a single float

* rearrange the eta_function integrals for readability and maintainability

gefux commented:
I've (with some efford) further simplified the eta_function computation. In particular I've ...

- replaced the previous default method with the cos() / sin() weighted version.
- simplified the integration parameters to be `epsrel`, `subdiv_limit`, and `omega_tau_threshold`. `omega_tau_threshold` is simply a `float`.
- separated the `eta_function()` into the `_eta_function_matsubara()` and `_eta_function()`.
- simplified the `_eta_function()` for readability and a shallower logic.

From my perspective this could be merged into the main branch and published with a minor version bump (i.e 0.5.1).
@emile-cochin and @piperfw: Let me know if you think we can go ahead, or if there are any more changes needed.

piperfw commented:
Thanks @gefux, I have been looking at the signatures of all the functions in `bath_correlations.py` and I think they are reasonable.

Ah, did you accidentally remove the new part of the parameters tutorial? I was about to suggest that be added...

gefux commented:
Oh yes. I'll add the tutorial comments back in this evening.

piperfw commented:
Thanks, feel free to remove or amend the '(see below)' comment I added to the notebook.

Another thing, I believe we now have three 'esprel' parameters: `epsrel` (TempoParameters), `liouvillian_epsrel` (System) and `epsrel` (Bath).

What do you think about renaming the last to `bath_epsrel` to avoid confusion? And similarly prefix the `integration_parameters` dictionary in the config.

gefux committed:
Reorganize integration parameters

gefux commented:
Thank you for that idea, @piperfw. I agree that the names should be somewhat consistent. I've gone over them and made some choices that seemed reasonable to me. I've changed `subdiv_limit` to `liouvillian_subdiv_limit` in `TempoParameters` and kept `epsrel`, `subdiv_limit`, `omega_tau_threshold` in the `CustomSD` class. That's becaus in `TempoParameters` it is necessary to note to what integral `subdiv_limit` referes, while not so in `CustomSD`. Also, if we continued down this road we'd have an `eta_custom_sd_epsrel` in `CustomSD.eta_function()` and an `correlation_custom_sd_epsrel` in `CustomSD.correlation()`, etc.

Remembering the Zen of Python: "Flat is better than nested." ... I've reverted to simply having two or three integration parameters, no dicts.

Also, the integration parameters for the spectral density classes are now set with the constructor. When a specific function, such as `CustomSD.correlation_2d_integral()` is called, one can either explicitly set the integration parameters or they default back to what has been set in the constructor.

Importantly TEMPO is now using the `epsrel` defined internally in the `CorrelationBase` class when calculating the influence functionals, and not the unrelated `epsrel` for the SVD.

Finally, I added back in a modified version of @emile-cochin 's parameters tutorial part.

These changes have now, in fact, become breaking changes. However, they are still so minor that I think we'd get away with them in a v0.5.1 ;)

piperfw commented:
Thanks @gefux, that all sounds reasonable. I would like to review the changes once more when I'm back at my desktop this weekend, thereafter think it is good to go

piperfw committed:
Parameter names for contraction functions

piperfw commented:
Sorry for the delay, I couldn't find time last weekend.

I've brought the parameters for the PT Tempo compute function in line with the new names. That's the only thing I could find. Tests and linter pass. @gefux please go ahead a merge if you're happy!

gefux commented:
@piperfw: thank you for checking and making some final corrections.
@emile-cochin: Thank you for this PR and the enhancement of OQuPy that comes with it. I am sorry for the rediculous delay in reviewing and merging this.
